src/Agent.Listener/Agent.cs

```diff
@@ -35,10 +34,16 @@ public override void Initialize(IHostContext hostContext)
 
         public async Task<int> ExecuteCommand(CommandSettings command)
         {
+            AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(Agent).GetTypeInfo().Assembly);
             try
             {
                 _inConfigStage = true;
+                _completedCommand.Reset();
                 _term.CancelKeyPress += CtrlCHandler;
+
+                //register a SIGTERM handler
+                loadContext.Unloading += Agent_Unloading;
```

Comment:
AgentService.js comment is unrelated. This change is not specific for any OS -> anytime the assembly is unloaded, we try to do a clean exit (if one has not been initiated already). I found that the assembly unload event is fired on Linux, but it does not work on OSX.